### DISCUSSION TITLE
Enable audio loopback devices

### DISCUSSION THREAD

**Discussion by aeantipov** [helgoboss/helgobox#1190] [Ideas]
Reaper now natively implement audio loopback devices. The main benefit is this allows a user to use reaper standard recording arming and monitoring for capturing output of a virtual output as opposed to using receive, which always sends signal to the output.
(https://reaper.blog/2022/10/669-update/) 
Using this (along with MIDI loopback) should in principle provide an alternative to #999 .
As an example, I can have a setup like this
![image](https://github.com/user-attachments/assets/3a09cb87-f863-4df6-bf9d-08162ef0f06a)
with potentially a virtual instrument on "Ext HW" track that sends it output to "Aux Loopback1" (here set to not record, but monitor only) 
The routing would look like this
![image](https://github.com/user-attachments/assets/bf754b2b-f097-4a99-a6ce-cc4eecab705e)

And playtime would see this with track 1 using "Aux Loopback1" input.
![image](https://github.com/user-attachments/assets/a967b22c-9ef1-4b53-b20b-a804cc5eef21)

It looks like it "almost works" - playtime in fact sees the signal, but says the input is unsupported.

**Comment by helgoboss** [comment]
> It looks like it "almost works" - playtime in fact sees the signal, but says the input is unsupported.

Oh! That's an important hint. Nice. It's possible that there's really not much effort necessary to make this happen. Will look into that.

**Comment by helgoboss** [comment]
Yes, this works! Awesome! Really glad you pointed me to audio loopback devices. I still think #999 is important, but this is a good and quick workaround.

Will be in 2.16.5.

**Reply by aeantipov** [reply]
Wonderful, that's great to hear. Hope it helps with bringing it faster to the point where people will adopt it to produce music.

**Reply by helgoboss** [reply]
I'm confident it will. I thought I would need to add #999 before the final release. But with audio loopbacks, I think this can move to the after-release phase.

**Reply by helgoboss** [reply]
2.16.5 is released. Feel free to test it.

**Reply by aeantipov** [reply]
Thanks! Do you know how long until it makes into reapack? Mine only shows 2.16.4 after synchronizing.

**Reply by helgoboss** [reply]
Sorry, had to take the release back. Will make another attempt later today

**Reply by helgoboss** [reply]
Available now in 2.16.5.

**Reply by aeantipov** [reply]
Thanks for the update! I am actively trying it out. It does work indeed. One minor tidbit 
Reaper somehow made a name clash - there are "loopback" devices and "aux loopback" devices. 
It looks like the official support is mainly the former ones, even though the latter ones is what advertised in the post linked at the top. 
**TLDR: Right now playtime only ~works~ shows correctly with loopback but not aux loopback.** 
It is minor in the sense, that despite what's written in reaper settings
![image](https://github.com/user-attachments/assets/b5c7a0b5-c490-4b65-a574-34e91e0b2225)
Loopback devices do actually work in the same project, and so indeed - this should be the supported way moving forward. 
However if supporting "Aux loopback" is an easy fix, I suggest allowing them too. 
To check it yourself, go to reaper.ini and do 
```
[reaper]
rearoute_loopback=8 # or whatever number
```
Then you'll see "aux_loopback" appearing in inputs. 
Right now playtime with "loopback" looks normal (track 1) but looks funky with aux loopback (track 2)
![image](https://github.com/user-attachments/assets/2fe930fb-d582-420f-bc6d-6fe7bee2c870)

**Reply by helgoboss** [reply]
Ah, interesting. That's what Justin meant by Aux Loopbacks. Yes, should be easy to allow support for those as well. Let me see.

**Reply by aeantipov** [reply]
I was trying to search on the exact difference between the two - AI totally outdid me.

## Reaper Loopback Outputs Comparison
In REAPER, both Aux Loopback and Loopback serve as virtual audio devices for routing audio between applications. The main difference lies in their functionality and usage:

### Loopback: This feature creates a virtual audio device that captures the output of REAPER and makes it available as an input to REAPER itself. Loopback is useful for:

- Recording REAPER’s output (e.g., for video syncing or podcasting)
- Monitoring REAPER’s output (e.g., for live performances or streaming)
- Creating a feedback loop for testing or debugging purposes
### Aux Loopback: This feature extends the Loopback functionality by allowing you to create multiple virtual audio devices, each with its own settings and routing options. Aux Loopback is useful for:

- Combining multiple audio sources (e.g., microphone, Skype, browser audio) into a single virtual device
- Routing these combined sources to different destinations (e.g., screen recording software, streaming software)
- Providing more flexibility and customization for complex audio routing scenarios

In summary:
Loopback is a single virtual device for capturing REAPER’s output, suitable for basic recording and monitoring tasks.
Aux Loopback is a more advanced feature, offering multiple virtual devices for combining and routing multiple audio sources, ideal for complex audio workflows.
When choosing between Aux Loopback and Loopback, consider the specific requirements of your project or workflow. If you need to record or monitor REAPER’s output only, Loopback might suffice. However, if you require more advanced audio routing and combination capabilities, Aux Loopback is the better choice.

**Reply by aeantipov** [reply]
I am not sure though it makes sense, anything you can do with one, you can do with the other.
Latency is also exactly the same
![image](https://github.com/user-attachments/assets/d6ecf858-5f5c-4929-b8d8-51af77d8c966)
![image](https://github.com/user-attachments/assets/33462d83-a749-4f62-85c5-552dd49ddd97)
